### ISSUE TITLE
metrics.accuracy_thresh() returns incorrect results

### ISSUE THREAD

**Issue by goerlitz** [fastai/fastai#2619]
**Describe the bug**

`metrics.accuracy_thresh()` does not return the correct accuracy - the calculated values seem to be too high (at least for the examples I checked). This accuracy metric is used in the [Planet Amazon Dataset](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson3-planet.ipynb) notebook and I have used it for my own multi-label problem where I double checked the accuracy of the results returned by `get_preds()`. The [metric documentation](https://docs.fast.ai/metrics.html#accuracy_thresh) mentions *one-hot-encoded targets* and *multiclassification problem* but the examples and tests do not included any real multi-class or multi-label examples. It seems to me that `metrics.accuracy_thresh()` has been broken since the beginning.

**Provide your installation details**

```text
=== Software === 
python       : 3.6.9
fastai       : 1.0.61
fastprogress : 0.2.3
torch        : 1.5.1
torch cuda   : 10.2 / is **Not available** 

=== Hardware === 
No GPUs available 

=== Environment === 
platform     : Linux-4.15.0-112-generic-x86_64-with-LinuxMint-19-tara
distro       : #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020
conda env    : Unknown
python       : /home/***/venv/bin/python3
sys.path     : /usr/lib/python36.zip
/usr/lib/python3.6
/usr/lib/python3.6/lib-dynload

/home/***/venv/lib/python3.6/site-packages
/home/***/venv/lib/python3.6/site-packages/IPython/extensions
/home/goerlitz/.ipython
no supported gpus found on this system
```

**To Reproduce**

This gist includes a reproducible example: https://gist.github.com/goerlitz/fb1230ee820406c5a9681249345fce3f

**Expected behavior**

computed accuracy should be correct.

**Screenshots**

**Additional context**

For multi-label predictions there are different possible ways how to compute the accuracy because a prediction may only match partially the target. I've included the computation of the accuracy for exact match and partial match in the gist. The way to calculate it is based on https://stats.stackexchange.com/questions/12702/what-are-the-measure-for-accuracy-of-multilabel-data

**Comment by jph00** [comment]
I believe the current approach is correct. In your first example in your gist, the right answer isn't 0.5, since that's ignoring all the correctly predicted "0"s.

**Comment by goerlitz** [comment]
Hi Jeremy,

could you explain why you want to include "all" predicted "0"s?

Sure, accuracy is basically (sum(TP) + sum(TN)) / total predictions for binary classification, but for multi-class and multi-label this way of calculating accuracy does not work anymore, mainly because we have just one "1" for the correct class and many "0"s for all other classes.

So here is a simple example. Assume we have a multi-class problem with 100 distinct classes (e.g. animals, plants, cars etc.).


A minimal (validation) dataset could have 100 observations - one for each of the 100 classes:

```
y_true = torch.eye(100,100)
```

A very naiive baseline model may always predict the same class (for all of the 100 observations):

```
y_pred = torch.cat([torch.ones(100,1),     # first class always 1.0
                    torch.zeros(100,99)],  # other classes always 0.0
                   dim=1)
```

This would leave us with only **one out of 100** observations having its class predicted correctly.

And here is what we get as results from the different accuracy implementations:

1. fastai
```
accuracy_thresh(y_pred, y_true, thresh=0.5, sigmoid=False)
tensor(0.9802)
```

2. sklearn
```
from sklearn.metrics import accuracy_score
accuracy_score(y_true.numpy(), y_pred.numpy())
0.01
```

3. my proposal
```
accuracy_multi_exact(y_pred, y_true, thresh=0.5, sigmoid=False)
tensor(0.0100)
```

In fact, for 100 classes the accuracy we would get when using `accuracy_thresh` would never be lower than 98% no matter how many observations we have in the validation dataset. This does not seem right.

**Comment by goerlitz** [comment]
It seems that a lot of time is spent on fastai2 these days. But if I'm not mistaken the same bug is in fastai2 too.

In fact the examples in my gist are correct - maybe they are too much simplified. I could actually make them more understandable if that helps.

So I'm wondering if this bug will be reopened or if it would be better to create a pull request?

**Comment by kinoute** [comment]
@goerlitz Do you have any news regarding this? I'm working on a multi-label classifier and struggle to understand why my `accuracy_multi` metric (I'm using FastAI2) is always greater than 99% no matter what. Since I'm working with more than 200 different labels, I highly doubt this is that accurate. Or I am missing the purpose of this metric.
